### PR TITLE
Fixes for torch cross_entropy function

### PR DESCRIPTION
**Pull request by hirwa-nshuti**
Hi @jkeane508 
Referring to issue #3043 
This PR is proposed to solve it and can complete issue #2751 Which is for the implementation of the torch cross_entropy function.

Thanks

### CODE REVIEW

File: ivy_tests/test_ivy/test_frontends/test_torch/test_loss_functions.py [hunk 1]
```diff
@@ -4,57 +4,84 @@
...
+@handle_cmd_line_args
 @given(
     dtype_and_input=helpers.dtype_and_values(
         available_dtypes=tuple(
             set(ivy_np.valid_float_dtypes).intersection(
                 set(ivy_torch.valid_float_dtypes)
             )
         ),
-        min_value=0,
+        min_value=1.0013580322265625e-05,
```
Comment (jkeane508) [inline ivy_tests/test_ivy/test_frontends/test_torch/test_loss_functions.py:21]: We can remove this line

File: ivy_tests/test_ivy/test_frontends/test_torch/test_loss_functions.py [hunk 2]
```diff
@@ -4,57 +4,84 @@
...
         max_value=1,
```
Comment (jkeane508) [inline ivy_tests/test_ivy/test_frontends/test_torch/test_loss_functions.py:22]: And this one

File: ivy_tests/test_ivy/test_frontends/test_torch/test_loss_functions.py [hunk 3]
```diff
@@ -4,57 +4,84 @@
...
+        max_dim_size=1,
     ),
     dtype_and_target=helpers.dtype_and_values(
         available_dtypes=tuple(
             set(ivy_np.valid_float_dtypes).intersection(
                 set(ivy_torch.valid_float_dtypes)
             )
         ),
-        min_value=1.0013580322265625e-05,
+        min_value=0,
```
Comment (jkeane508) [inline ivy_tests/test_ivy/test_frontends/test_torch/test_loss_functions.py:37]: we can remove

File: ivy_tests/test_ivy/test_frontends/test_torch/test_loss_functions.py [hunk 4]
```diff
@@ -4,57 +4,84 @@
...
         max_value=1,
```
Comment (jkeane508) [inline ivy_tests/test_ivy/test_frontends/test_torch/test_loss_functions.py:38]: we can remove also

File: ivy_tests/test_ivy/test_frontends/test_torch/test_loss_functions.py [hunk 5]
```diff
@@ -4,57 +4,84 @@
...
+        min_dim_size=1,
+        max_dim_size=1,
+    ),
+    dtype_and_weights=helpers.dtype_and_values(
+        available_dtypes=tuple(
+            set(ivy_np.valid_float_dtypes).intersection(
+                set(ivy_torch.valid_float_dtypes)
+            )
+        ),
+        min_value=1.0013580322265625e-05,
```
Comment (jkeane508) [inline ivy_tests/test_ivy/test_frontends/test_torch/test_loss_functions.py:51]: We can remove this

File: ivy_tests/test_ivy/test_frontends/test_torch/test_loss_functions.py [hunk 6]
```diff
@@ -4,57 +4,84 @@
...
+        max_value=1.0,
```
Comment (jkeane508) [inline ivy_tests/test_ivy/test_frontends/test_torch/test_loss_functions.py:52]: And this

File: ivy_tests/test_ivy/test_frontends/test_torch/test_loss_functions.py [hunk 7]
```diff
@@ -4,57 +4,84 @@
...
 import ivy.functional.backends.numpy as ivy_np
 import ivy.functional.backends.torch as ivy_torch
 
 
 # cross_entropy
+@handle_cmd_line_args
 @given(
     dtype_and_input=helpers.dtype_and_values(
         available_dtypes=tuple(
             set(ivy_np.valid_float_dtypes).intersection(
```
Comment (jkeane508) [inline ivy_tests/test_ivy/test_frontends/test_torch/test_loss_functions.py:17]: We can probably accept ints for this one

File: ivy/functional/frontends/torch/loss_functions.py [hunk 8]
```diff
@@ -2,6 +2,42 @@
 import ivy
 
 
+def _get_reduction_method(reduction, to_reduce):
+    if reduction == "none":
+        ret = to_reduce
+    elif reduction == "mean":
+        ret = ivy.mean(to_reduce)
+    elif reduction == "elementwise_mean":
```
Comment (jkeane508) [inline ivy/functional/frontends/torch/loss_functions.py:10]: This is never tested, why is it in if its not a documented parameter

File: ivy/functional/frontends/torch/loss_functions.py [hunk 9]
```diff
@@ -12,7 +48,12 @@ def cross_entropy(
     reduction="mean",
     label_smoothing=0.0,
 ):
-    return ivy.cross_entropy(input, target)
+    input = ivy.softmax(input)
```
Comment (jkeane508) [inline ivy/functional/frontends/torch/loss_functions.py:51]: Why do we softmax here?

**Review by jkeane508 — CHANGES_REQUESTED** [review]
Couple of changes needed here. I understand this is a difficult implementation. Just need some clarification of some lines below :)

File: ivy_tests/test_ivy/test_frontends/test_torch/test_loss_functions.py (hunk 7 above)
Comment (hirwa-nshuti) [reply ivy_tests/test_ivy/test_frontends/test_torch/test_loss_functions.py:17]: No, because the `torch.nn.functional.cross_entropy` is a log softmax and only accepts floats and doubles as inputs.

File: ivy/functional/frontends/torch/loss_functions.py (hunk 8 above)
Comment (hirwa-nshuti) [reply ivy/functional/frontends/torch/loss_functions.py]: This one is deprecated in Pytorch and am going to remove it from here.

File: ivy/functional/frontends/torch/loss_functions.py (hunk 9 above)
Comment (hirwa-nshuti) [reply ivy/functional/frontends/torch/loss_functions.py:49]: The `cross_entropy` function applies LogSoftmax and then NLLLoss. As it is in the implementation details here 
[CROSSENTROPYLOSS](https://pytorch.org/docs/stable/generated/torch.nn.CrossEntropyLoss.html#torch.nn.CrossEntropyLoss)
This softmax here is for converting the input to a normalized probability distribution. then we apply our ivy `cross_entropy` loss function.

File: ivy_tests/test_ivy/test_frontends/test_torch/test_loss_functions.py (hunk 6 above)
Comment (hirwa-nshuti) [reply ivy_tests/test_ivy/test_frontends/test_torch/test_loss_functions.py]: 👍

File: ivy_tests/test_ivy/test_frontends/test_torch/test_loss_functions.py (hunk 5 above)
Comment (hirwa-nshuti) [reply ivy_tests/test_ivy/test_frontends/test_torch/test_loss_functions.py]: 👍

File: ivy_tests/test_ivy/test_frontends/test_torch/test_loss_functions.py (hunk 4 above)
Comment (hirwa-nshuti) [reply ivy_tests/test_ivy/test_frontends/test_torch/test_loss_functions.py]: 👍

File: ivy_tests/test_ivy/test_frontends/test_torch/test_loss_functions.py (hunk 3 above)
Comment (hirwa-nshuti) [reply ivy_tests/test_ivy/test_frontends/test_torch/test_loss_functions.py]: 👍

File: ivy_tests/test_ivy/test_frontends/test_torch/test_loss_functions.py (hunk 2 above)
Comment (hirwa-nshuti) [reply ivy_tests/test_ivy/test_frontends/test_torch/test_loss_functions.py]: 👍

File: ivy_tests/test_ivy/test_frontends/test_torch/test_loss_functions.py (hunk 1 above)
Comment (hirwa-nshuti) [reply ivy_tests/test_ivy/test_frontends/test_torch/test_loss_functions.py]: 👍

File: ivy_tests/test_ivy/test_frontends/test_torch/test_loss_functions.py (hunk 7 above)
Comment (jkeane508) [reply ivy_tests/test_ivy/test_frontends/test_torch/test_loss_functions.py:17]: Ok sounds good

File: ivy/functional/frontends/torch/loss_functions.py (hunk 9 above)
Comment (jkeane508) [reply ivy/functional/frontends/torch/loss_functions.py:49]: Ok that makes sense

File: ivy_tests/test_ivy/test_frontends/test_torch/test_loss_functions.py (hunk 3 above)
Comment (jkeane508) [reply ivy_tests/test_ivy/test_frontends/test_torch/test_loss_functions.py]: Ok sounds good

File: ivy_tests/test_ivy/test_frontends/test_torch/test_loss_functions.py [hunk 10]
```diff
@@ -4,57 +4,74 @@
...
         ),
-        min_value=1.0013580322265625e-05,
-        max_value=1,
         allow_inf=False,
-        exclude_min=True,
-        exclude_max=True,
-        min_num_dims=1,
-        max_num_dims=1,
-        min_dim_size=2,
+        min_num_dims=2,
```
Comment (jkeane508) [inline ivy_tests/test_ivy/test_frontends/test_torch/test_loss_functions.py:36]: Same here

**Review by jkeane508 — CHANGES_REQUESTED** [review]
I have just read through the docs, I think the shape generation for the testing here will require helper functions to test this properly. I don't think 2 dims is sufficient.

File: ivy_tests/test_ivy/test_frontends/test_torch/test_loss_functions.py [hunk 11]
```diff
@@ -4,57 +4,74 @@
...
     dtype_and_input=helpers.dtype_and_values(
         available_dtypes=tuple(
             set(ivy_np.valid_float_dtypes).intersection(
                 set(ivy_torch.valid_float_dtypes)
             )
         ),
-        min_value=0,
-        max_value=1,
         allow_inf=False,
         min_num_dims=2,
```
[the remark this answers is not in the document]
Comment (hirwa-nshuti) [reply ivy_tests/test_ivy/test_frontends/test_torch/test_loss_functions.py:22]: I actually read it also from the documentation that the input tensor is of (C) or (minibatch, C) or minibatch, C, d1, d2, ..., dk) for k >= 1, and the weight should always be a 1D tensor,  of size C, I do agree that there needs to be a helper function for drawing these testing values.

File: ivy_tests/test_ivy/test_frontends/test_torch/test_loss_functions.py [hunk 12]
```diff
@@ -64,4 +81,10 @@ def test_torch_cross_entropy(
         fn_tree="nn.functional.cross_entropy",
         input=np.asarray(input, dtype=inputs_dtype),
         target=np.asarray(target, dtype=target_dtype),
+        weight=np.asarray(weights, dtype=weights_dtype),
+        size_average=size_average,
+        ignore_index=-100,
```
[the remark this answers is not in the document]
Comment (hirwa-nshuti) [reply ivy_tests/test_ivy/test_frontends/test_torch/test_loss_functions.py:88]: Used to mask some given labels but as I haven't implemented it I'll need to remove it from the test function and also figured out that the `label_smoothing` acts as the epsilon value from `ivy.cross_entropy`, going to add it to the function implementation.

File: ivy_tests/test_ivy/test_frontends/test_torch/test_loss_functions.py (hunk 11 above)
[the remark this answers is not in the document]
Comment (hirwa-nshuti) [reply ivy_tests/test_ivy/test_frontends/test_torch/test_loss_functions.py:22]: I was reading on some issues and found that the target Tensor has to be a 1D array for cross_entropy, so I only limited the dims for it as well as the weights. if yet the input tensor should be of as many dims as possible. I think the helper function is probably not needed here.
